### ISSUE TITLE
Flyouts do not adapt to match Compact Density

### ISSUE THREAD

**Issue by mdtauk** [microsoft/microsoft-ui-xaml#1148]
**Describe the bug**
ComboBox flyouts are larger when you use a touch screen to make it drop down, compared to using a cursor.
Calendar, Date and Time Pickers, do not, and perhaps they should.

The design of the Date and Time picker centres the flyout above the control, and so the selection matches the 32px height, so it appears seamless, as if the whole control is expanding out.

When Compact Density is used, the control becomes 24px, but the flyout isn't affected.  This is not what you would expect, and appears to be a bug.

If some of the proposals ( #918 #735 #905 ) are considered, then this disconnect between the flyout and the control it expands from, will look even more stark.

**Steps to reproduce the bug**
* Add controls to your page.
* Apply the Compact Density to your app.xaml or to the controls.
* Tap with a touch screen, or click with your mouse, to open the flyout on the control.
* Observe the fact the flyouts do not correctly use the compact metrics.

**Expected behavior**
![image](https://user-images.githubusercontent.com/7389110/62748615-2d802180-ba51-11e9-994f-b0a27fb53f75.png)
_This is an illustrative example of expected behaviour_

**Screenshots**

**Version Info**
All versions with support for Compact Density

| Device form factor | Saw the problem? |
| :-------------------- | :------------------- |
| Desktop                 | Yes  |
| Surface Go                   | Yes |

**Additional context**
Related GitHub Issues: #925 #1079

**Comment by jevansaks** [comment]
I think this was a known limitation of the density work and will require WinUI 3.0 fixes to address. Is that right, @chrisglein @LucasHaines ?

**Comment by chrisglein** [comment]
@licanhua did the work on compact. I know there were open issues with the dictionary and API that had the feature left in preview. But I don't remember what the limitations were. For app.xaml level customization I would expect this to work. @licanhua was this part of the known issues with where we left compact?

**Comment by licanhua** [comment]
I will try to list the restrictions I remeber one by one here:
1. Xaml framework treats the keys differently based on if it's defined in generic.xaml or not.
For example, compact/density may not work well in below situation or may have down-level problem.
Because the TextBox/ComboBox and other controls existing in WUX is defined in os generic.xaml, and it has high priority in resource lookup. And styles in XamlControlsResources is skipped.

 ```
 <Application.Resources>
      <ResourceDictionary>
            <ResourceDictionary.MergedDictionaries>
                <local:ItemTemplates />
                <XamlControlsResources xmlns="using:Microsoft.UI.Xaml.Controls"/> <- CombBox/TextBox in MUX is ignored
            </ResourceDictionary.MergedDictionaries>

            <Style TargetType="TextBox">
                <Setter Property="PlaceholderForeground" Value="{ThemeResource TextControlPlaceholderForeground}" />  high priority
            </Style>

            <Style TargetType="ComboBox">
                <Setter Property="PlaceholderForeground" Value="{ThemeResource ComboBoxPlaceHolderForeground}" />   high priority
            </Style>
</ResourceDictionary>
```

**Comment by licanhua** [comment]
2. In below example, Color Red is actual selected.

```
Dictionary1.xaml
<ResourceDictionary
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation" 
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:MergedDict">
    <SolidColorBrush x:Key="MyButtonBackground" Color="Green" />
 
</ResourceDictionary>
 
Dictionary2.xaml:
<ResourceDictionary
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation" 
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:MergedDict">
    <SolidColorBrush x:Key="MyButtonBackground" Color="Red" />
 
    <Style TargetType="Button">
        <Setter Property="Background" Value="{ThemeResource MyButtonBackground}" />
    </Style>
</ResourceDictionary>
 
Dictionary3.xaml
<ResourceDictionary
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation" 
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:MergedDict">
    <SolidColorBrush x:Key="MyButtonBackground" Color="Green" />
 
</ResourceDictionary>
 
App.xaml
<Application
    x:Class="MergedDict.App"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:MergedDict">
    <Application.Resources>
        <ResourceDictionary>
            <ResourceDictionary.MergedDictionaries>
 
                <ResourceDictionary Source="ms-appx:///Dictionary1.xaml" />
                <ResourceDictionary Source="ms-appx:///Dictionary2.xaml" />
                <ResourceDictionary Source="ms-appx:///Dictionary3.xaml" />
 
            </ResourceDictionary.MergedDictionaries>
            <SolidColorBrush x:Key="MyButtonBackground" Color="Yellow" />
 
        </ResourceDictionary>
    </Application.Resources>
</Application>
```

Here is the explain from Alan:
You have to look at it from the perspective of (a) what the framework knows about (i.e. what global dictionaries exist), (b) what the parser knows about while reading the current XBF, (c) and how this affects what the resource resolution walk encounters. The reason this doesn't work is a fairly straightforward consequence of how the XAML framework operates and how XAML markup maps to code.

The parser *does not* have lookahead capability. 
Implication: forward-references to static/theme resources are forbidden. 
Corollary: a reference to a static/theme resource will completely ignore an override that you specify later in the file.

The parser does not assign objects to a property until the element end tag is encountered. 
Implication: '<ResourceDictionary Source="Dictionary2.xaml"></ResourceDictionary>' has no idea that it is inside Application.Resources.MergedDictionaries until after "</ResourceDictionary>" is encountered, which is after Dictionary2.xaml has been run through the parser. 
Implication: the ResourceDictionary defined by Dictionary2.xaml *has no logical parent* at the time of parse because it can't be assigned to the (eventual parent) MergedDictionaries collection until after it has finished initializing, i.e. the parser has finished loading Dictionary2.xaml
Corollary: the other dictionaries specified in Application.Resources.MergedDictionaries are invisible to Dictionary2.xaml
Implication: from the perspective of the framework, Application.Resources *does not exist* while App.xaml is being loaded because you can't assign a value to a variable before that value has finished initializing
Corollary: the resource resolution walk for a static/theme resource reference inside Dictionary2.xaml will only encounter resources that were defined earlier in Dictionary2.xaml or in the platform generic.xaml

The code equivalent of assigning your App.xaml to Application.Resources would be something like this:
```
Application.Resources = null;
var appXaml = new ResourceDictionary();
var dictionary1 = new ResourceDictionary();
dictionary1.Source = new Uri("Dictionary1.xaml"); // loads the specified markup via black box
appXaml.MergedDictionaries.Add(dictionary1);

var dictionary2 = new ResourceDictionary();
dictionary2.Source = new Uri("Dictionary2.xaml"); // loads the specified markup via black box
appXaml.MergedDictionaries.Add(dictionary2);

var dictionary3 = new ResourceDictionary();
dictionary3.Source = new Uri("Dictionary3.xaml"); // loads the specified markup via black box
appXaml.MergedDictionaries.Add(dictionary3);

Application.Resources = appXaml;
```
It should be clear from the above why what you're trying to do doesn't work; while Dictionary2.xaml is being loaded, Application.Resources is still null dictionary3 doesn't even exist yet, and dictionary2 does not have a parent because it has yet to be added to the appXaml.MergedDictionaries collection. As a result, static/theme resource references inside Dictionary2.xaml literally have no access to any resources defined in Dictionary1.xaml or Dictionary3.xaml.

**Comment by licanhua** [comment]
3. some height is hard coded to 44 before RS4. Most likely part of this bug is because of this. The flyout may look differently in RS5+.
for example: DatePickerFlyoutPresenterItemHeight.
Expect this problem is resolved automatically in winui 3.0.

**Comment by mdtauk** [comment]
This order of inheritence sounds like it is going to be something to change when WinUI 3.0 is released.  The OS should not override any control templates, but instead just provide the colours, theme choice, and user settings, for XAML to consume - but hopefully everything else bar the core windowing, would be owned by Microsoft.UI.Xaml

**Comment by jevansaks** [comment]
Yes, this is a temporary problem so we should do the best we can for compact density and the new visual updates -- but it may mean that there are some visual updates we can't make or that we can't make easy for developers to opt out. Hopefully we won't be in this limbo for too long.

**Comment by mdtauk** [comment]
As long as a solution comes eventually, I don't think it is a huge enough bug to worry about right now.  But it may be worth collating similar framework issues that may be holding Microsoft and others back, which could be fixed once XAML is decoupled from the OS cycle.

**Comment by mdtauk** [comment]
This is still an issue

**Comment by niels9001** [comment]
@mdtauk Compact density styles aren't really s thing anymore in WinUI 3. Should we close this issue?

**Comment by mdtauk** [comment]
> [@mdtauk](https://github.com/mdtauk) Compact density styles aren't really s thing anymore in WinUI 3. Should we close this issue?

I think its ok to close it.
